pkg/sql/sqlutil/internal_executor.go

```diff
@@ -196,16 +207,32 @@ type InternalRows interface {
 	Types() colinfo.ResultColumns
 }
 
-// SessionBoundInternalExecutorFactory is a function that produces a "session
-// bound" internal executor.
-type SessionBoundInternalExecutorFactory func(
+// InternalExecutorFactory is a function that produces an internal executor.
+type InternalExecutorFactory func(
 	context.Context, *sessiondata.SessionData,
 ) InternalExecutor
 
+// WithoutTxn is to run SQL statements without a txn context.
+// Note that in this mode, each execution with the internal executor runs its
+// own descriptor collection, schema change job collection and transaction state
+// machine.
+func (f InternalExecutorFactory) WithoutTxn(
+	ctx context.Context, run func(ctx context.Context, ie InternalExecutor) error,
+) error {
+	ie := f(ctx, nil /* SessionData */)
+	return run(ctx, ie)
+}
```

Comment:
In what case would one want to use this? Or rather, why would anybody ever want to invoke `f` and not use this?

Reply:
> why would anybody ever want to invoke f and not use this?

I'm guessing you were trying to say "not invoke f but use this"? If this is what you wanted to say, yeah I agree that this function cannot "disallow" users to invoke the factory themselves to create an internal executor. The intent was to make it consistent that an internal executor instance should only live in the scope of certain functions (such as `ief.WithoutTxn(), cf.WithTxn, planner.ExecEx()` and etc. But I agree that this solution cannot truly eliminate the motivation for undesired usage.